### PR TITLE
ADR 012: Layers

### PR DESCRIPTION
**Pull request by opqdonut**
# Definition of Done / Review checklist

## Documentation
- [x] ADR for major architectural decisions or experiments

### CODE REVIEW

File: docs/architecture/012-services.md [hunk 1]
```diff
@@ -17,13 +18,20 @@ was reached.
 
 ## Decision
 
-Each service represents the public API for a specific part of REMS.
-This public API can be used to implement the HTTP API, but also from
-other parts of the system.
+Each service represents the public API for a specific part of REMS that is easy to call
+and handles dependencies to other items:
+- resource, workflow, ... structure
```
Comment (opqdonut) [inline docs/architecture/012-services.md:23]: This list could be a bit clearer. Are these examples of services? Perhaps mention the namespaces as well?

Comment (opqdonut) [reply docs/architecture/012-services.md]: better now, thanks

File: docs/architecture/012-layers.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+This ADR tries to document the current state of the REMS application
+architecture for the backend.
+
+## Layers
+
+### The API Layer
+
+The API layer takes in a HTTP request and transforms it into ideally
+one call of the service layer.
+
```
Comment (Macroz) [inline docs/architecture/012-layers.md:14]: How about adding responsibilities?
- coarse-grained access control through `:roles` annotation
- controlled schemas
- swagger documentation

File: docs/architecture/012-layers.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+### The DB Layer
+
+The purpose of the DB layer is to handle serializing and deserializing
+data into the database. A single namespace in the DB layer should
+handle one concept (for example `rems.db.licenses`). The DB layer can
+also contain simple domain logic (for example `rems.db.attachments`
+has checking for allowed attachment types). Namespaces in the DB layer
+should not depend on each other. Depending on the common
+`rems.db.core` namespace is fine though.
+
```
Comment (Macroz) [inline docs/architecture/012-layers.md:47]: How about adding something about how the db-internal keys should be translated into model keys?

Comment (Macroz) [reply docs/architecture/012-layers.md:47]: I think it should happen here and the keys I mean are mostly old keys or `endt` etc. Also organization is just a string id column and it should be a `{:organization/id xxx}` structure after load?

Comment (opqdonut) [reply docs/architecture/012-layers.md:47]: :+1:

File: docs/architecture/012-layers.md (hunk 2 above)
Comment (opqdonut) [reply docs/architecture/012-layers.md:14]: :+1:

**Review by Macroz — APPROVED** [review]

**Review by foxlynx — APPROVED** [review]

**Pull request merged**